### PR TITLE
bump ibc to v4.2.0 with ibcfee 

### PR DESCRIPTION
**Pull request by yaruwangway**

### CODE REVIEW

**Comment by faddat** [comment]
might want this in v8?

**Comment by faddat** [comment]
Also, SL has put out a v4 pfm branch

**Comment by yaruwangway** [comment]
related issues:
[#55 - The router module is not an IBC middleware](https://github.com/strangelove-ventures/packet-forward-middleware/issues/55)
[#56 - fix: implement the ibc middleware interface](https://github.com/strangelove-ventures/packet-forward-middleware/pull/56)

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -277,19 +279,51 @@ func NewAppKeeper(
...
+	var transferStack porttypes.IBCModule
+	transferStack = transfer.NewIBCModule(appKeepers.TransferKeeper)
+	transferStack = ibcfee.NewIBCMiddleware(transferStack, appKeepers.IBCFeeKeeper)
+	transferStack = router.NewIBCMiddleware(
+		transferStack,
+		appKeepers.RouterKeeper,
+		0,
+		routerkeeper.DefaultForwardTransferPacketTimeoutTimestamp,
+		routerkeeper.DefaultRefundTransferPacketTimeoutTimestamp,
+	)
```
Comment (damiannolan) [inline app/keepers/keepers.go]: I think you will need to swap these two lines. 

From what I'm looking at, the desired flow is:
`SendPacket`:  transfer -> router -> fee -> ibc core
`RecvPacket`: ibc core -> fee -> router -> transfer

```suggestion
	var transferStack porttypes.IBCModule
	transferStack = transfer.NewIBCModule(appKeepers.TransferKeeper)
	transferStack = router.NewIBCMiddleware(
		transferStack,
		appKeepers.RouterKeeper,
		0,
		routerkeeper.DefaultForwardTransferPacketTimeoutTimestamp,
		routerkeeper.DefaultRefundTransferPacketTimeoutTimestamp,
	)
	transferStack = ibcfee.NewIBCMiddleware(transferStack, appKeepers.IBCFeeKeeper)
```

What you're seeing currently is ibc core passing an ics29 fee acknowledgement to the packet forward router module. And the router failing to unmarshal the ack [here](https://github.com/strangelove-ventures/packet-forward-middleware/blob/v4.0.0/router/ibc_middleware.go#L254-L257). 

If ibc core passes the acknowledgement firstly to ics29 fee, it should successfully [unmarshal the incentivized ack](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/ibc_middleware.go#L252-L255) and then [pass `ack.AppAcknowledgement` to the next handler in the call stack](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/ibc_middleware.go#L292).

**Comment by glnro** [comment]
would not merge this into main before ics, maybe label do not merge or convert to draft

File: app/keepers/keepers.go (hunk 1 above)
Comment (yaruwangway) [reply app/keepers/keepers.go]: I see, thank you, Damian!👍

**Comment by yaruwangway** [comment]
> might want this in v8?

Hi this is for v10. v9 is still using ibc-go v3.4

File: tests/e2e/e2e_ibc_test.go [hunk 2]
```diff
@@ -34,7 +34,9 @@ type PacketMetadata struct {
 	Forward *ForwardMetadata `json:"forward"`
 }
 
-func (s *IntegrationTestSuite) runIBCRelayer() {
+// setupIBCRelayer creates a Hermes relayer container and establishes
+// connections and channels between the two test chains.
+func (s *IntegrationTestSuite) setupIBCRelayer() {
```
Comment (MSalopek) [inline tests/e2e/e2e_ibc_test.go:39]: Spliting this func into 2 parts (setup and start) allows us to establish connections/channels and run some state assertions before we actually clear any packets with hermes.

**Pull request closed without merging**